[JSC] Refactor String slice/substring helper functions

Gumichocopengin8 opened the pull request:
#### 3b290941ead8a92a10c8385c5f4a20beb9c05f7d
<pre>
[JSC] Refactor String slice/substring helper functions
<a href="https://bugs.webkit.org/show_bug.cgi?id=300870">https://bugs.webkit.org/show_bug.cgi?id=300870</a>

Reviewed by Darin Adler.

This patch refactors String slice/substring helper functions
for readability.

* Source/JavaScriptCore/runtime/StringPrototypeInlines.h:
(JSC::extractSubstringOffsets):

Canonical link: <a href="https://commits.webkit.org/301790@main">https://commits.webkit.org/301790@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3b8ee6efe055160056932c4216a0f54093dacf4a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126956 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46592 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37582 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133960 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78531 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47209 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55120 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96604 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64633 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129904 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37777 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113554 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77116 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36639 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31721 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77353 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/118993 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107614 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32042 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136485 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/125419 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53612 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41279 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/105122 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54115 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109919 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104813 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26751 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50331 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28670 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/51091 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53544 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/59416 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/158457 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52784 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39638 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/56118 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54544 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/runtime/StringPrototypeInlines.h:
```diff
@@ -69,6 +69,7 @@ ALWAYS_INLINE std::tuple<int32_t, int32_t> extractSliceOffsets(int32_t length, i
 }
 
 template<typename NumberType>
+requires std::integral<NumberType> || std::floating_point<NumberType>
```
I think it would be better to use a concept here rather than `requires`. Unfortunately, the library doesn’t have the concept already. So close, though! So we can add the concept to WTF in `StdLibExtras.h` or right here:
```suggestion
template<typename T> concept Arithmetic = std::is_arithmetic_v<T>;

template<Arithmetic NumberType>
```

Gumichocopengin8 replied:
Updated

darinadler commented on Source/JavaScriptCore/runtime/StringPrototypeInlines.h:
```diff
@@ -48,7 +48,7 @@ inline Structure* StringPrototype::createStructure(VM& vm, JSGlobalObject* globa
 }
 
-ALWAYS_INLINE std::tuple<int32_t, int32_t> extractSliceOffsets(int32_t length, int32_t startValue, std::optional<int32_t> endValue)
+ALWAYS_INLINE std::pair<int32_t, int32_t> extractSliceOffsets(int32_t length, int32_t startValue, std::optional<int32_t> endValue)
```
Change seems OK, but why is `std::pair` better than `std::tuple`?

Gumichocopengin8 replied:
The main reason is semantics. It only returns 2 values and we don't expect it to return 3 or more values in the future. Semantically `std::pair` fits better than `std::tuple` here, especially for readers of the code.
`std::pair` has standard layout for simple types (predictable memory layout), whereas `std::tuple` does not. 
```cpp
std::cout << (std::is_standard_layout_v<std::tuple<int, int>>) << std::endl; // false
std::cout << (std::is_standard_layout_v<std::pair<int, int>>) << std::endl; // true
```

darinadler replied:
I don't find either of these arguments compelling, personally, although I don't care passionately about it.

For something returning two values, `pair` does not seem to be clearer than `tuple`. In fact I think it goes the other way around. A `tuple` means there is more than one thing and imparts no special meaning to the number of items. But `pair` seems to imply there is something meaningful about the fact that there are specifically two things, and in these cases there is not.

That memory layout doesn't seem important at all. We don't need a guaranteed layout nor do we benefit from it.

But I won't stand in the way of this change at this time. Seems arbitrary either way without compelling arguments.

Constellation replied:
Yeah, I think `tuple` is better than `pair` as well. The result is currently using tuple just to returning multiple values, and there is a chance that we can start returning more things. So changing it to `pair` is misleading since it implies that value must be paired, but in fact, it is not so much.

Constellation replied:
Also, this is `ALWAYS_INLINE`, and the expected behavior is just,
```
auto [x, y] = func();
```
So, layout does not matter much.

Gumichocopengin8 replied:
Updated. Reverted `std::pair`

darinadler commented on Source/JavaScriptCore/runtime/StringPrototypeInlines.h:
```diff
@@ -89,12 +90,10 @@ ALWAYS_INLINE JSString* stringSlice(JSGlobalObject* globalObject, VM& vm, JSStri
 }
 
-ALWAYS_INLINE std::tuple<int32_t, int32_t> extractSubstringOffsets(int32_t length, int32_t startValue, std::optional<int32_t> endValue)
+ALWAYS_INLINE std::pair<int32_t, int32_t> extractSubstringOffsets(int32_t length, int32_t startValue, std::optional<int32_t> endValue)
```
Change seems OK, but why is `std::pair` better than `std::tuple`?

darinadler commented on Source/JavaScriptCore/runtime/StringPrototypeInlines.h:
```diff
@@ -89,12 +90,10 @@ ALWAYS_INLINE JSString* stringSlice(JSGlobalObject* globalObject, VM& vm, JSStri
-    if (endValue)
-        end = std::min<int32_t>(std::max<int32_t>(endValue.value(), 0), length);
+    int32_t start = std::clamp<int32_t>(startValue, 0, length);
+    int32_t end = std::clamp<int32_t>(endValue.value_or(length), 0, length);
```
I wish we could always use `std::clamp` instead of `std::min` combined with `std::max`. But it’s not _always_ OK, because if the lower bound is higher than the upper bound the behavior of `std::clamp` is undefined (I was shocked to discover this), and in most cases it returns the upper bound in this case, whereas `std::min` and `std::max` can be used in a way that guarantees the result is the lower bounds in that case, which is typically the more useful behavior in my experience.

But in this case, since `length` is guaranteed to be >= 0, it’s fine to use `std::clamp`.

I suggest we not write out the types explicitly, because everything here is `int32_t`, including the 0, since in practice int is always `int32_t` in all the configurations we support.

```suggestion
    int32_t start = std::clamp(startValue, 0, length);
    int32_t end = std::clamp(endValue.value_or(length), 0, length);
```

Gumichocopengin8 replied:
Updated

Gumichocopengin8 committed:
[JSC] Refactor String slice/substring helper functions
https://bugs.webkit.org/show_bug.cgi?id=300870

Reviewed by Darin Adler.

This patch refactors String slice/substring helper functions
for readability.

* Source/JavaScriptCore/runtime/StringPrototypeInlines.h:
(JSC::extractSubstringOffsets):

Canonical link: https://commits.webkit.org/301790@main
